pkg/ottl/contexts/ottllog/log.go

```diff
@@ -208,7 +220,10 @@ func (pep *pathExpressionParser) parsePath(path ottl.Path[TransformContext]) (ot
 		} else {
 			return accessSpanID(), nil
 		}
+	case "schemaURL":
```

Comment:
I believe we'll need to add special logic in the `resource` and `instrumentaiton_scope` cases.  If `path.Name() == resource` and `path.Next().Name() == "schema_url" then call `accessResourceSchemaURL`.

Something like that.

Reply:
Yes

Reply:
Sorry for jumping in a bit late here, but I think we should adjust course just a bit. I think what we have now is on the right track, but we should implement something like the following type in the `internal` package:

```golang
type SchemaURLItem interface {
    SchemaUrl() string
    SetSchemaUrl(v string)
}
```

Then make methods like this on the `TransformContext` type that implement corresponding methods on the `ResourceContext`/`InstrumentationScopeContext` interfaces like this:

```golang
func (tCtx TransformContext) GetResourceSchemaURLItem() internal.SchemaURLItem {
	return tCtx.resourceLogs
}
```
The resource and scope path parsers could then just call each of these methods to implement the necessary getters and setters similar to how we're already doing it in ottllog.

As it is, I'm worried that re-implementing this for each package will produce a lot of unnecessary duplication.

Reply:
Ya sorry I haven't gotten to spend a lot time thinking about the details of this PR. Thanks for steering it correctly.

Reply:
> Sorry for jumping in a bit late here, but I think we should adjust course just a bit. I think what we have now is on the right track, but we should implement something like the following type in the `internal` package:
> 
> ```go
> type SchemaURLItem interface {
>     SchemaUrl() string
>     SetSchemaUrl(v string)
> }
> ```

@evan-bradley yeah this make sense since with different ottl packages, we would probably have to write a lot of similar code. So this sort of abstraction should work for our case. 

I am bit new to the project structure, so could you point me to the module inside `internal` in which this interface could fall. 

> Then make methods like this on the `TransformContext` type that implement corresponding methods on the `ResourceContext`/`InstrumentationScopeContext` interfaces like this:
> 
> ```go
> func (tCtx TransformContext) GetResourceSchemaURLItem() internal.SchemaURLItem {
> 	return tCtx.resourceLogs
> }
> ```
> 
> The resource and scope path parsers could then just call each of these methods to implement the necessary getters and setters similar to how we're already doing it in ottllog.
> 
> As it is, I'm worried that re-implementing this for each package will produce a lot of unnecessary duplication.

But I still feel that this abstraction would not have an effect on the NewTransformContext Function which we need to change for each module in ottl.

```go
func NewTransformContext(logRecord plog.LogRecord, instrumentationScope pcommon.InstrumentationScope, resource pcommon.Resource, scopeLogs plog.ScopeLogs, resourceLogs plog.ResourceLogs) TransformContext {
	return TransformContext{
		logRecord:            logRecord,
		instrumentationScope: instrumentationScope,
		resource:             resource,
		cache:                pcommon.NewMap(),
		scopeLogs:            scopeLogs,
		resourceLogs:         resourceLogs,
	}
}
```

For instance, for ottllog we would still have to pass in resourceLogs and scopeLogs. For the path parser as well would still have to call 2 separate functions each for access schemaURL using resourceLogs and scopeLogs.

Reply:
@evan-bradley @TylerHelmuth any thoughts here ?

Reply:
Hey @kernelpanic77, sorry, I wasn't able to look at this yesterday.

>any thoughts here on where I should add schemaURLItem interface ?

Add it to a new file in `pkg/ottl/contexts/internal` called `schema.go`.

>For instance, for ottllog we would still have to pass in resourceLogs and scopeLogs. For the path parser as well would still have to call 2 separate functions each for access schemaURL using resourceLogs and scopeLogs.

You're right that we would need to add getters for the schemaURL for the resource and scope levels of the TransformContext and would need to pass e.g. resourceLogs and scopeLogs into ottllog's `NewTransformContext` function. We would need to implement the getter in each of `internal/resource.go` and `internal/scope.go`, then they could just handle the `schema_url` subpath under the `resource` and `scope` paths respectively.

I realize that's pretty wordy, let me know if any of that isn't clear.

Reply:
> You're right that we would need to add getters for the schemaURL for the resource and scope levels of the TransformContext and would need to pass e.g. resourceLogs and scopeLogs into ottllog's `NewTransformContext` function. We would need to implement the getter in each of `internal/resource.go` and `internal/scope.go`, then they could just handle the `schema_url` subpath under the `resource` and `scope` paths respectively.
@evan-bradley  I am not sure if i understood this correctly, since implementing the accessSchemaURL function for `internal/resource.go` and `internal/scope.go` would require access to resoure/scope Log objects. I am not sure how we can access that.

Reply:
@evan-bradley  Have a look at my latest commit and let me know if this is what you were expecting. 

I am not sure though how the `parsePath` function inside `ottllog/log.go` would change. If could add in a commit to this branch, it would be really helpful.

Reply:
@evan-bradley does the latest commit contain what you were looking for ?

Reply:
@kernelpanic77 Sorry for the delay. I think this is in the right direction, but I'd make a few changes. I left a review to help explain what I'm thinking.